File: docs/snippets/all/archetypes/gaussian_splats3d_simple.cpp
// Log a few gaussian splats.

#include <rerun.hpp>

int main(int argc, char* argv[]) {
    const auto rec = rerun::RecordingStream("rerun_example_gaussian_splats3d");
    rec.spawn().exit_on_failure();

    // 15 view-dependent RGB coefficients (degrees 1-3) per splat, coefficient-major:
    std::array<std::array<float, 3>, 15> red_sh, green_sh, blue_sh;
    red_sh.fill({0.5f, 0.0f, 0.0f});
    green_sh.fill({0.0f, 0.5f, 0.0f});
    blue_sh.fill({0.0f, 0.0f, 0.5f});

    rec.log(
        "gaussians",
        rerun::GaussianSplats3D(
            {{0.0f, 0.0f, 0.0f}, {2.0f, 0.0f, 0.0f}, {4.0f, 0.0f, 0.0f}}
        )
            .with_scales(
                {{1.0f, 0.5f, 0.25f}, {0.5f, 1.0f, 0.5f}, {0.25f, 0.5f, 1.0f}}
            )
            .with_quaternions({
                rerun::Quaternion::IDENTITY,
                // 45 degrees around Z
                rerun::Quaternion::from_xyzw(0.0f, 0.0f, 0.382683f, 0.923880f),
                rerun::Quaternion::IDENTITY,
            })
            .with_colors({
                rerun::Rgba32(255, 0, 0, 128),
                rerun::Rgba32(0, 255, 0, 200),
                rerun::Rgba32(0, 0, 255, 255),
            })
            .with_sh_coefficients({
                rerun::datatypes::SphericalHarmonics3Rgb(red_sh),
                rerun::datatypes::SphericalHarmonics3Rgb(green_sh),
                rerun::datatypes::SphericalHarmonics3Rgb(blue_sh),
            })
    );
}

File: docs/snippets/all/archetypes/gaussian_splats3d_simple.rs
//! Log a few gaussian splats.

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec =
        rerun::RecordingStreamBuilder::new("rerun_example_gaussian_splats3d")
            .spawn()?;

    rec.log(
        "gaussians",
        &rerun::GaussianSplats3D::new([
            (0.0, 0.0, 0.0),
            (2.0, 0.0, 0.0),
            (4.0, 0.0, 0.0),
        ])
        .with_scales([(1.0, 0.5, 0.25), (0.5, 1.0, 0.5), (0.25, 0.5, 1.0)])
        .with_quaternions([
            rerun::Quaternion::IDENTITY,
            rerun::Quaternion::from_xyzw([0.0, 0.0, 0.382683, 0.923880]), // 45 degrees around Z
            rerun::Quaternion::IDENTITY,
        ])
        .with_colors([
            rerun::Color::from_unmultiplied_rgba(255, 0, 0, 128),
            rerun::Color::from_unmultiplied_rgba(0, 255, 0, 200),
            rerun::Color::from_unmultiplied_rgba(0, 0, 255, 255),
        ])
        // 15 view-dependent RGB coefficients (degrees 1-3) per splat, coefficient-major:
        .with_sh_coefficients([
            [[0.5, 0.0, 0.0]; 15],
            [[0.0, 0.5, 0.0]; 15],
            [[0.0, 0.0, 0.5]; 15],
        ]),
    )?;

    Ok(())
}
